cpp/src/arrow/dataset/file_parquet.cc

```diff
@@ -523,6 +523,10 @@ Result<RecordBatchGenerator> ParquetFileFormat::ScanBatchesAsync(
     ARROW_ASSIGN_OR_RAISE(row_groups, parquet_fragment->FilterRowGroups(options->filter));
     pre_filtered = true;
     if (row_groups.empty()) return MakeEmptyGenerator<std::shared_ptr<RecordBatch>>();
+    if (options->start_offset != kDefaultStartOffset) {
+      ARROW_ASSIGN_OR_RAISE(row_groups,
```

Comment:
when `row_groups` is not empty, they means the row-groups after applying `options->filter`. Here it will recreate `row_groups`, which may contain rowgroups not in `FilterRowGroups`

Reply:
correct

Reply:
@mapleFU  updated, please take another look

Reply:
This look good to me now!